providers/dnsimple/dnsimpleProvider.go

```diff
@@ -0,0 +1,314 @@
+	corrections := []*models.Correction{}
+
+	records, err := c.getRecords(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	var actual []*models.RecordConfig
+	for _, r := range records {
+		fqdn := r.Name + "." + dc.Name
+		fqdn = strings.TrimLeft(fqdn, ".")
+		actual = append(actual, &models.RecordConfig{
+			NameFQDN: fqdn,
+			Type:     r.Type,
+			Target:   r.Content,
+			TTL:      uint32(r.TTL),
+			Priority: uint16(r.Priority),
+			Original: r,
+		})
+	}
+
+	differ := diff.New(dc)
+	_, create, delete, modify := differ.IncrementalDiff(actual)
+
+	for _, del := range delete {
+		rec := del.Existing.Original.(dnsimpleapi.ZoneRecord)
+		corrections = append(corrections, &models.Correction{
+			Msg: del.String(),
+			F:   c.deleteRecordFunc(rec.ID, dc.Name),
+		})
+	}
+
+	for _, cre := range create {
+		rec := cre.Desired
+		corrections = append(corrections, &models.Correction{
+			Msg: cre.String(),
+			F:   c.createRecordFunc(rec, dc.Name),
+		})
+	}
+
+	for _, mod := range modify {
+		old := mod.Existing.Original.(dnsimpleapi.ZoneRecord)
+		new := mod.Desired
+		corrections = append(corrections, &models.Correction{
+			Msg: mod.String(),
+			F:   c.updateRecordFunc(&old, new, dc.Name),
+		})
+	}
+
+	return corrections, nil
+}
+
+func (c *DnsimpleApi) GetRegistrarCorrections(dc *models.DomainConfig) ([]*models.Correction, error) {
+	corrections := []*models.Correction{}
+
+	nameServers, err := c.getNameservers(dc.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	actual := strings.Join(nameServers, ",")
+
+	expectedSet := []string{}
+	for _, ns := range dc.Nameservers {
+		expectedSet = append(expectedSet, ns.Name)
+	}
+	sort.Strings(expectedSet)
+	expected := strings.Join(expectedSet, ",")
+
+	if actual != expected {
+		return []*models.Correction{
+			{
+				Msg: fmt.Sprintf("Update nameservers %s -> %s", actual, expected),
+				F:   c.updateNameserversFunc(expectedSet, dc.Name),
+			},
+		}, nil
+	}
+
+	return corrections, nil
+}
+
+// DNSimple calls
+
+func (c *DnsimpleApi) getClient() *dnsimpleapi.Client {
+	client := dnsimpleapi.NewClient(dnsimpleapi.NewOauthTokenCredentials(c.AccountToken))
+	if c.BaseURL != "" {
+		client.BaseURL = c.BaseURL
+	}
+	return client
+}
+
+func (c *DnsimpleApi) getAccountId() (string, error) {
+	if c.accountId == "" {
+		client := c.getClient()
+		whoamiResponse, err := client.Identity.Whoami()
+		if err != nil {
+			return "", err
+		}
+
+		c.accountId = strconv.Itoa(whoamiResponse.Data.Account.ID)
+	}
+	return c.accountId, nil
+}
+
+func (c *DnsimpleApi) getRecords(domainName string) ([]dnsimpleapi.ZoneRecord, error) {
+	client := c.getClient()
+
+	accountId, err := c.getAccountId()
+	if err != nil {
+		return nil, err
+	}
+
+	recordsResponse, err := client.Zones.ListRecords(accountId, domainName, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return recordsResponse.Data, nil
+}
+
+// Returns the name server names that should be used. If the domain is registered
```

Comment:
Are both registrar and DNS provider expected to use this same function? If so, I'll need to separate those implementations.